components/proxy/src/main/kotlin/com/hotels/styx/routing/db/StyxObjectStore.kt

```diff
@@ -15,28 +15,100 @@
  */
 package com.hotels.styx.routing.db;
 
-import com.hotels.styx.api.configuration.ObjectStore
+import com.hotels.styx.api.configuration.ObjectStoreSnapshot
+import org.pcollections.HashTreePMap
+import org.pcollections.PMap
+import org.reactivestreams.Publisher
+import reactor.core.publisher.Flux
 import java.util.Optional
-import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent.Executors
+import java.util.concurrent.atomic.AtomicReference
+import java.util.function.Consumer
 
 /**
  * Styx Route Database.
  */
-class StyxObjectStore<T> : ObjectStore<T> {
-    private val objects = ConcurrentHashMap<String, Record<T>>();
+class StyxObjectStore<T> : ObjectStoreSnapshot<T> {
+    private val objects: AtomicReference<PMap<String, Record<T>>> = AtomicReference(HashTreePMap.empty())
+    private val watchers = CopyOnWriteArrayList<ChangeWatcher<T>>()
+    private val executor = Executors.newSingleThreadExecutor()
 
-    override fun get(name: String?): Optional<T> {
-        return Optional.ofNullable(objects.get(name))
+    override fun get(name: String): Optional<T> {
+        return Optional.ofNullable(objects().get(name))
```

Comment:
Would you prefer to use the indexing operator? objects()[name] .We could also do objects() -  instead of minus() , etc. but all these seem a bit 'magical' compared to the more verbose method invocations.

Reply:
Hi @dvlato is the indexing operator identical in its behaviour when a key is not found? Specifically, will it return a null value or does it just throw?

Reply:
The indexing operator is translated by an invocation to a get() function, so it will be completely identical, yes. See here how operator overloading works in Kotlin: https://kotlinlang.org/docs/reference/operator-overloading.html